src/smtencoding/FStar.SMTEncoding.Encode.fst

```diff
@@ -1000,6 +1000,547 @@ let encode_top_level_let :
+          let _, _, _, u_formals = TcTerm.tc_binders env_tps formals in
+          let inj = BU.for_all (fun u_formal -> u_leq_u_k u_formal) u_formals in                  
+          if inj
+          then (
+            match (SS.compress t).n with
+            | Tm_type u -> 
+            (* retain injectivity for parameters that are type functions
+                from small universes (i.e., all formals are smaller than the constructed type)
+                to a universe <= the universe of the constructed type.
+                See BugBoxInjectivity.fst *)
+              u_leq_u_k u
+            | _ ->
+              false
+          )
+          else (
+            false
+          )
+
+        )
+    in
+    let is_injective_on_params = List.forall2 tp_ok tps us in
+    if Env.debug env.tcenv <| Options.Other "SMTEncoding"
+    then BU.print2 "%s injectivity for %s\n"
+                (if is_injective_on_params then "YES" else "NO")
+                (Ident.string_of_lid t);
+    is_injective_on_params
+
+
+let encode_sig_inductive (is_injective_on_params:bool) (env:env_t) (se:sigelt)
+: decls_t * env_t
+= let Sig_inductive_typ { lid=t; us=universe_names; params=tps; t=k; ds=datas} = se.sigel in 
+  let t_lid = t in
+  let tcenv = env.tcenv in
+  let quals = se.sigquals in
+  let is_logical = quals |> BU.for_some (function Logic | Assumption -> true | _ -> false) in
+  let constructor_or_logic_type_decl (c:constructor_t) =
+    if is_logical
+    then [Term.DeclFun(c.constr_name, c.constr_fields |> List.map (fun f -> f.field_sort), Term_sort, None)]
+    else constructor_to_decl (Ident.range_of_lid t) c in
+  let inversion_axioms env tapp vars =
+    if datas |> BU.for_some (fun l -> Env.try_lookup_lid env.tcenv l |> Option.isNone) //Q: Why would this happen?
+    then []
+    else (
+      let xxsym, xx = fresh_fvar env.current_module_name "x" Term_sort in
+      let data_ax, decls =
+        datas |>
+        List.fold_left
+          (fun (out, decls) l ->
+            let is_l = mk_data_tester env l xx in
+            let inversion_case, decls' =
+              if is_injective_on_params
+              || Options.ext_getv "compat:injectivity" <> ""
+              then (
+                let _, data_t = Env.lookup_datacon env.tcenv l in
+                let args, res = U.arrow_formals data_t in
+                let indices = res |> U.head_and_args_full |> snd in
+                let env = args |> List.fold_left
+                    (fun env ({binder_bv=x}) -> push_term_var env x (mkApp(mk_term_projector_name l x, [xx])))
+                    env in
+                let indices, decls' = encode_args indices env in
+                if List.length indices <> List.length vars
+                then failwith "Impossible";
+                let eqs = List.map2 (fun v a -> mkEq(mkFreeV v, a)) vars indices in
+                mkAnd(is_l, mk_and_l eqs), decls'
+              )
+              else is_l, []
```

Comment:
If it's not injective, then just generate the simple inversion that provides a disjunction on the constructors of the inductive